accepted/2020/minimum-os-version/minimum-os-version.md

```diff
@@ -0,0 +1,246 @@
+
+## NuGet Manifest
+
+When generating a NuGet package using MSBuild, any platform minimum version constraints expressed in the project file will be listed in the `.nuspec`  manifest in a `<platformInfo>` child of the `<metadata>` element. This element and all of its children are optional. They expose the same information as the assembly annotation attributes, but avoid the need for consumers to read the assembly metadata. This could potentially be validated and/or automated by NuGet..
+
+Within the `<platformInfo>` there are child elements corresponding to the TFM short names (used by NuGet as the folder name) of the reference assemblies in the NuGet. If the element represents a platform-specific TFM (e.g. `net5-ios14.0` or `net6-android9.0`) then it must have a `minimumVersion` attribute representing the minimum platform version for that platform:
+
+```xml
+<metadata>
+    <platformInfo>
+    <net5-ios15.0 minimumVersion="13.0" />
+    <net5-mac10.15 minimumVersion="10.11" />
+    </platformInfo>
+</metadata>
+```
+
+A reference assembly for a platform-neutral TFM (e.g. `net5` or `net6`) may still need to express platform-specific minimum versions if the implementations are platform-specific (aka ‘bait and switch’). As before, the element matches the TFM of the reference assembly, but this time there is no `minimumVersion`. Instead there are sub-elements that provide `minimumVersion` values for platforms using the `TargetPlatformIdentifier` of those platforms to identify them:
+
+```xml
+<metadata>
+    <platformInfo>
+    <net6>
+        <ios minimumVersion="13.0">
+        <mac minimumVersion="11.0">
+    </net6>
+    </platformInfo>
+</metadata>
+```
+
+Using these mechanisms, a NuGet package can declare minimum platform versions for any or all of its reference assemblies.
+
+# Availability Inheritance
+
+Members and nested types that do not have explicit availability annotations will inherit the values from their containing types. Types that do not have `Introduced` annotations will be treated as if they were introduced in the minimum platform version of the assembly itself.
+
+# IDE Experience
+
+Quick Info and IntelliSense tooltips will display the minimum platform version for any member with a minimum platform version higher than the current project’s minimum platform version.
+
+We may also want to consider some kind of sorting, grouping or colorization based hints in the IntelliSense list.
+
+# Analyzers
+
+There will be a Roslyn analyzer that will squiggle calls to any method or property with a minimum platform version higher than the current project’s minimum platform version, unless either of the following is true:
+
+
+- the call is guarded by a `CheckOS` or `CheckOSVersion` call against a version that is greater than or equal to the member’s minimum platform version
+- the caller member has a minimum platform version that is equal to or higher than the callee
```

Comment:
There's potential to have analyzers to ensure availability attributes are logical - but it brings some special cases.

E.g. when a new type is introduced in a hierarchy, which is not a breaking change, you can end up with

```csharp
[Introduced (.ios, 8,0)]
class NSFoo : NSObject {
   string Id {get;}
   string Name {get;}
}
```
becoming
```csharp
[Introduced (.ios, 14,0)]
class NSMiddle : NSObject {
   [Introduced (.ios, 8,0)]
   string Id {get;}
}

[Introduced (.ios, 8,0)]
class NSFoo : NSMiddle {
   string Name {get;}
}
```

and it's easy to think there's a mistake in `Middle` since a member was available well before it's type. While you can't create `NSMiddle` on iOS8 you can still use `NSFoo.Middle` (without warnings) on iOS8.

Reply:
We can use [Windows](https://docs.microsoft.com/en-us/windows/uwp/debug-test-perf/version-adaptive-code)'s existing version and api tagging or we could go with existing tags which are already there to expose them as such in a more .NET/WinMD way. Existing tools could be easily supported if we did follow latter idea.

> Try to use ideas as much as from the Windows platform teams. They are relevant to all other platforms out there.